dissect/target/filesystems/jffs.py

```diff
@@ -0,0 +1,120 @@
+from typing import BinaryIO, Iterator, Optional
+
+from dissect.jffs import c_jffs2, jffs2
+
+from dissect.target.exceptions import (
+    FileNotFoundError,
+    FilesystemError,
+    IsADirectoryError,
+    NotADirectoryError,
+    NotASymlinkError,
+)
+from dissect.target.filesystem import Filesystem, FilesystemEntry
+from dissect.target.helpers import fsutil
+
+
+class JFFSFilesystem(Filesystem):
+    __fstype__ = "jffs"
+
+    def __init__(self, fh: BinaryIO, *args, **kwargs):
+        super().__init__(fh, *args, **kwargs)
+        self.jffs2 = jffs2.JFFS2(fh)
+
+    @staticmethod
+    def _detect(fh: BinaryIO) -> bool:
+        block = fh.read(2)
+        magic = int.from_bytes(block[0:2], "little")
+
+        return magic in (
+            c_jffs2.c_jffs2.JFFS2_MAGIC_BITMASK,
+            c_jffs2.c_jffs2.JFFS2_OLD_MAGIC_BITMASK,
+        )
+
+    def get(self, path: str) -> FilesystemEntry:
+        return JFFSFilesystemEntry(self, path, self._get_node(path))
+
+    def _get_node(self, path: str, node: Optional[jffs2.INode] = None):
+        try:
+            return self.jffs2.get(path, node)
+        except jffs2.FileNotFoundError as e:
+            raise FileNotFoundError(path, cause=e)
+        except jffs2.NotADirectoryError as e:
+            raise NotADirectoryError(path, cause=e)
+        except jffs2.NotASymlinkError as e:
+            raise NotASymlinkError(path, cause=e)
+        except jffs2.Error as e:
+            raise FileNotFoundError(path, cause=e)
+
+
+class JFFSFilesystemEntry(FilesystemEntry):
+    def get(self, path: str) -> FilesystemEntry:
+        entry_path = fsutil.join(self.path, path, alt_separator=self.fs.alt_separator)
+        entry = self.fs._get_node(path, self.entry)
+        return JFFSFilesystemEntry(self.fs, entry_path, entry)
+
+    def open(self) -> BinaryIO:
+        if self.is_dir():
+            raise IsADirectoryError(self.path)
+        return self._resolve().entry.open()
+
+    def iterdir(self) -> Iterator[str]:
+        if not self.is_dir():
+            raise NotADirectoryError(self.path)
+
+        if self.is_symlink():
+            for name, entry in self.readlink_ext().iterdir():
+                yield name
+        else:
+            for name, entry in self.entry.iterdir():
+                yield name
+
+    def scandir(self) -> Iterator[FilesystemEntry]:
+        for name in self.iterdir():
+            entry_path = fsutil.join(self.path, name, alt_separator=self.fs.alt_separator)
+            entry = self.fs.get(entry_path).entry
```

Comment:
Can you tweak the logic of `iterdir()` and `scandir()` to be closer to this:
https://github.com/fox-it/dissect.target/blob/main/dissect/target/filesystems/ffs.py#L62-L83

The logic here being that by iterating over the `INode.iterdir()` items, you already have the `INode` and don't need to do another lookup with `self.fs.get()` and can instead re-use that object.

Reply:
That makes sense, added in [80c5706](https://github.com/fox-it/dissect.target/pull/417/commits/80c570630c5e6db80c84255b409e96d47c9e2197).